### ISSUE TITLE
Question: how to handle paste without redraw

### ISSUE THREAD

**Issue by iphydf** [jtdaugherty/brick#362]
https://github.com/jtdaugherty/brick/blob/ec2f45f07e2d510a0ff9c48d982ac940f7a92401/programs/EditDemo.hs#L57

Every time a character is entered, the UI redraws, which takes significant time. Pasting 100 characters takes almost a second on my machine in non-optimised code (it's reasonably fast in optimised code, but it's still a waste of compute and I'd like to be able to load very large pastes into my application to support pasting base64-encoded images into it).

What's the recommended way to avoid redraws when the next event is e.g. less than 1ms after the current one?

Bracketed paste mode works nicely, but some terminals (e.g. GNU screen) don't support it.

Edit: it may take several seconds to paste 100 characters when the screen is somewhat full of text. This is reduced a lot when using the rendering cache. Is that the recommended way? It's still somewhat slow, but reasonable.

**Comment by iphydf** [comment]
Correction: it may take several seconds to paste 100 characters when the screen is somewhat full of text.

**Comment by jtdaugherty** [comment]
First, bracketed paste mode is by far the best way to go if you can use it. Second, if you can access the system clipboard (e.g. using [Hclip](https://hackage.haskell.org/package/Hclip)) then you can add a shortcut key to paste directly from that and draw the UI only once. If you can't use paste mode or Hclip, then you're right that some rendering cache management is likely going to be needed.

**Comment by iphydf** [comment]
One alternative I considered but so far didn't implement because it's super ugly is to never actually redraw on input events and instead start a very short timer of say 1ms to redraw later, and reset that timer on every input. Some applications like irssi seem to do something like that for resize (but not paste). I could try it as a proof of concept in my application but if this approach is reasonable/feasible I think it should rather be done by the framework.

I can't use bracketed paste in screen. It's quite low priority for me though because in screen sessions I'm never going to paste huge amounts of data, and cache plus ghc optimisation makes normal paste fast enough (so far). In irssi I don't paste large amounts of text either, so the use case isn't strong.

**Comment by jtdaugherty** [comment]
Okay, thanks for your update. For the approach you described, I guess I'm with you on its aesthetics. I agree that it would be ideal if the framework took responsibility for supporting that kind of thing, but I'd really like to hold off on what I'd view as a concession of defeat and see if your other work produces enough of a gain in performance. 😊

**Comment by iphydf** [comment]
Closing this for now, until it becomes a problem again. Caching is probably the answer.
